### DISCUSSION TITLE
Allowing File Access from Anywhere in Tauri v2

### DISCUSSION THREAD

**Discussion by maxbodin** [tauri-apps/tauri#11792] [Q&A]
# Allowing File Access from Anywhere in Tauri v2

## Description

Hello everyone, I'm encountering an issue while developing a Tauri v2 application. The goal of my app is to allow users to access **any type of file, regardless of its location on their computer**. However, I'm running into what seems to be strict restrictions imposed by Tauri on file access.

### Problem Details

Initially, I tried using the `@tauri-apps/plugin-dialog` to let users select files. However, the need to specify a `BaseDirectory` makes it overly restrictive. Even after updating the scopes of the `fs` plugin with "*", I was still unable to access files freely. 

I then attempted to implement a backend command to read files directly using the following Rust code:

```rust
#[tauri::command]
fn read_file(path: String) -> Response {
  let data = std::fs::read(path).unwrap();
  tauri::ipc::Response::new(data)
}
```

In the frontend, I invoked the command like this:

```typescript
const readFile = (path) => {
  invoke("read_file", { path }).then((message) => console.log(message));
};

readFile("/Users/BOB/Documents/test.png");
```

Unfortunately, this results in a PermissionDenied error:
```bash
thread 'main' panicked at src/lib.rs:11:34:
called `Result::unwrap()` on an `Err` value: Os { code: 1, kind: PermissionDenied, message: "Operation not permitted" }
```

### Question
1. Is there any way to bypass Tauri's restrictions to access files freely?
2. If such unrestricted file access isn't possible, is Tauri the wrong choice for this type of application?

### My Thoughts
While I understand the importance of security, I find these restrictions too limiting for many practical use cases. Users should be able to open files from anywhere on their system without running into PermissionDenied errors. The current design feels overly restrictive and makes me question how suitable Tauri is for building such applications.

If anyone has suggestions or workarounds, I would greatly appreciate your help!

**Comment by FabianLars** [comment]
> However, the need to specify a BaseDirectory makes it overly restrictive

There should be no api that _requires_ the use of BaseDirectory. It should always be optional unless i forgot some api here.

> Even after updating the scopes of the fs plugin with "*", I was still unable to access files freely.

Try `**/*`

> Unfortunately, this results in a PermissionDenied error:

This is an OS error, completely uncoupled from tauri's permission system. (Tauri cannot really affect/restrict the Rust side anyway)

The only thing i can think of on macOS, apart from a generally protected path (if that exists on macos), is the `Unrestricted File Access` privacy setting (idk how exactly it was called) which may have to be added to the Terminal in dev and not just the app.
Or a more simple thing, file permissions (the chmod thingy). did you try multiple files in multiple locations?

**Comment by QuentinWach** [comment]
@maxbodin, I have written about file management in Tauri 2.0 [here](https://quentinwach.com/blog/2024/11/26/files-in-tauri-v2.html). No `BaseDirectory` required. I hope it helps.

**Comment by na2axl** [comment]
Adding this permission helped me fix the issue:
```json
{
  "identifier": "fs:scope",
  "allow": [{ "path": "**/*" }]
}
```

**Reply by gesslar** [reply]
Put it where? I cannot for the life of me figure out where to put this setting? I'm trying everywhere in `tauri.conf.json`, but it is `permissions` is an illegal key no matter where I put it.

**Reply by FabianLars** [reply]
in the json file in `src-tauri/capabilities/` which looks something like this https://v2.tauri.app/plugin/file-system/#permissions

**Comment by haexhub** [comment]
A bit OT, but when we are talking about file access permissions. How is it possible / best practice to grant MANAGE_EXTERNAL_STORAGE on android systems?

**Comment by oojr** [comment]
@maxbodin this is the entitlement you need
```  
  <key>com.apple.security.temporary-exception.files.absolute-path.read-write</key>
  <array>
      <string>/</string>
  </array>
  ```
